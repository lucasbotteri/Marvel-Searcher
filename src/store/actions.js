import {
  SET_LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
  SET_COMICS,
  SET_LOADING_COMICS,
  SET_COMIC,
  SET_LOADING_COMIC,
  SAVE_FAV,
  SHOW_FAVS,
  REMOVE_FAV,
  HIDE_FAVS,
} from "./constants";
import {
  getCharacterByName,
  getRandomCharacter,
  getCharacterComics,
  getComic,
} from "../service";

export const setLoadingCharacters = () => ({
  type: SET_LOADING_CHARACTERS_LIST,
});

export const setCharacters = (characters) => ({
  type: SET_CHARACTERS_LIST,
  payload: characters,
});

export const setCharacterSelected = (id, name) => ({
  type: SET_CHARACTER_SELECTED,
  payload: { id, name },
});

export const hideModal = () => ({
  type: HIDE_MODAL,
});

export const setComics = (comics) => ({
  type: SET_COMICS,
  payload: comics,
});

export const setLoadingComics = () => ({
  type: SET_LOADING_COMICS,
});

export const setComic = (comics) => ({
  type: SET_COMIC,
  payload: comics,
});

export const setLoadingComic = () => ({
  type: SET_LOADING_COMIC,
});

export const saveFav = (character) => ({
  type: SAVE_FAV,
  payload: character,
});

export const showFavs = () => ({
  type: SHOW_FAVS,
});

export const hideFavs = () => ({
  type: HIDE_FAVS,
});

export const removeFav = (id) => ({
  type: REMOVE_FAV,
  payload: { id },
});

export const searchCharacter = async (dispatch, characterName) => {
  dispatch(setLoadingCharacters());
  const characters = await getCharacterByName(characterName);
  dispatch(setCharacters(characters));
};

export const searchRandomCharacter = async (dispatch) => {
  dispatch(setLoadingCharacters());
  const characters = await getRandomCharacter();
  dispatch(setCharacters(characters));
};

export const searchCharacterComics = async (
  dispatch,
  characterId,
  comicFilter
) => {
  dispatch(setLoadingComics());
  const comics = await getCharacterComics(characterId, comicFilter);
  dispatch(setComics(comics));
};

export const searchComic = async (dispatch, comicId) => {
  dispatch(setLoadingComic());
  const comic = await getComic(comicId);
  dispatch(setComic(comic));
};
