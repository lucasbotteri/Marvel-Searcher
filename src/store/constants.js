export const LOADING_CHARACTERS_LIST = "LOADING_CHARACTERS_LIST";
export const SET_CHARACTERS_LIST = "SET_CHARACTERS_LIST";
export const SET_CHARACTER_SELECTED = "SET_CHARACTER_SELECTED";
export const HIDE_MODAL = "HIDE_MODAL";
export const SET_COMICS = "SET_COMICS";
export const SET_LOADING_COMICS = "SET_LOADING_COMICS";
