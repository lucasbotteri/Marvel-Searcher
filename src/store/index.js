export { default as StoreProvider, StoreContext } from "./StoreProvider";
