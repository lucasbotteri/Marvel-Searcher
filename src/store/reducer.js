import {
  LOADING_CHARACTERS_LIST,
  SET_CHARACTERS_LIST,
  SET_CHARACTER_SELECTED,
  HIDE_MODAL,
} from "./constants";

const reducer = (state, action) => {
  switch (action.type) {
    case LOADING_CHARACTERS_LIST:
      return { ...state, characters: [], isCharacterListLoading: true };
    case SET_CHARACTERS_LIST:
      return {
        ...state,
        characters: action.payload,
        isCharacterListLoading: false,
        isShowingModal: false,
      };
    case SET_CHARACTER_SELECTED:
      return {
        ...state,
        characterSelected: action.payload,
        isShowingModal: true,
      };
    case HIDE_MODAL:
      return {
        ...state,
        isShowingModal: false,
      };
    default:
      throw new Error("Invalid Action");
  }
};

export default reducer;
