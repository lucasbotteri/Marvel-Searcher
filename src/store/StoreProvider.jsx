import React, { useReducer } from "react";
import PropTypes from "prop-types";
import reducer from "./reducer";
import * as actions from "./actions";

const INITAL_STATE = {
  characters: [],
  loadingCharacters: false,
  fetchedCharactersSuccessful: true,
};

export const StoreContext = React.createContext();

const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, INITAL_STATE);
  const value = {
    ...state,
    fetchCharacters: (data) => actions.fetchCharacters(dispatch, data),
  };
  return (
    <StoreContext.Provider value={value}>{children}</StoreContext.Provider>
  );
};

StoreProvider.propTypes = {
  children: PropTypes.element.isRequired,
};

export default StoreProvider;
