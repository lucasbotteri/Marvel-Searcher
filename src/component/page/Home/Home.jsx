import React, { useContext, useEffect } from "react";
import styled from "styled-components";
import CharacterList from "../../organism/CharacterList";
import CharacterModal from "../../organism/CharacterModal";
import ComicDetail from "../../molecule/ComicDetail";
import { StoreContext } from "../../../store";
import { useQuery } from "../../../hook";

const HomeWrapper = styled.main`
  background-color: ${(props) => props.theme.appBackgroundColor};
  box-sizing: border-box;
  height: 100%;
  padding-top: 90px;
  text-align: center;

  @media (max-width: 425) {
    margin: 0;
  }
`;

const Home = () => {
  const {
    searchRandomCharacter,
    searchCharacter,
    comic,
    characters,
    loadingContent,
    isShowingFavs,
    favs,
  } = useContext(StoreContext);

  const { character: queryParamCharacter } = useQuery();

  const charactersToShow = isShowingFavs ? favs : characters;

  const noResults = !loadingContent && !charactersToShow && !comic;

  useEffect(() => {
    if (queryParamCharacter) {
      searchCharacter(queryParamCharacter);
    } else {
      searchRandomCharacter();
    }
  }, [searchRandomCharacter, searchCharacter, queryParamCharacter]);

  return (
    <HomeWrapper>
      {loadingContent && "...loading"}
      {noResults && "No Results Found"}
      {charactersToShow && <CharacterList characters={charactersToShow} />}
      {comic && <ComicDetail comic={comic} />}
      <CharacterModal />
    </HomeWrapper>
  );
};

export default Home;
