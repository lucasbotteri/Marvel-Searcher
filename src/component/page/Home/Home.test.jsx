import React from "react";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import { StoreContext } from "../../../store";
import Home from "./Home";

const characters = [
  {
    id: 1011377,
    name: "Spider-Man (Takuya Yamashiro)",
    thumbnail:
      "http://i.annihil.us/u/prod/marvel/i/mg/f/50/4be86ad8ada17/portrait_fantastic.jpg",
  },
  {
    id: 1011010,
    name: "Spider-Man (Ultimate)",
    thumbnail:
      "http://i.annihil.us/u/prod/marvel/i/mg/3/50/531771b4e8c60/portrait_fantastic.jpg",
  },
];

const contextStore = {
  searchRandomCharacter: () => {},
  searchCharacters: () => {},
  comic: null,
  characters: null,
  loadingContent: false,
  isShowingFavs: false,
  favs: null,
};

const mountHome = (overrideStore, overrideEntry) => {
  return mount(
    <MemoryRouter initialEntries={[overrideEntry || "/"]}>
      <StoreContext.Provider value={{ ...contextStore, ...overrideStore }}>
        <Home characters={characters} />
      </StoreContext.Provider>
    </MemoryRouter>
  );
};

describe("Home", () => {
  it("Should render without crashing", () => {
    const wrapper = mountHome();
    expect(wrapper).toBeTruthy();
  });

  it("Should render loading", () => {
    const wrapper = mountHome({
      loadingContent: true,
    });
    expect(wrapper.containsMatchingElement("...loading")).toBeTruthy();
  });

  it("Should render no results", () => {
    const wrapper = mountHome();
    expect(wrapper.containsMatchingElement("No Results Found")).toBeTruthy();
  });

  it("Should call searchRandomCharacter", () => {
    const searchRandomCharacter = jest.fn();
    mountHome({
      searchRandomCharacter,
    });
    expect(searchRandomCharacter).toHaveBeenCalled();
  });

  it("Should call searchCharacter", () => {
    const searchCharacters = jest.fn();
    mountHome(
      {
        searchCharacters,
      },
      "/?character=spiderman"
    );
    expect(searchCharacters).toHaveBeenCalledWith(["spiderman"]);
  });

  it("Should match snapshot", () => {
    const wrapper = mountHome({ characters });
    expect(wrapper).toMatchSnapshot();
  });
});
