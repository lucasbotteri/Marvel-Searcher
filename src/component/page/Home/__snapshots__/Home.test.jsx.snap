// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should match snapshot 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      "/",
    ]
  }
  keyLength={0}
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Home
      characters={
        Array [
          Object {
            "id": 1011377,
            "name": "Spider-Man (Takuya Yamashiro)",
            "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/f/50/4be86ad8ada17/portrait_fantastic.jpg",
          },
          Object {
            "id": 1011010,
            "name": "Spider-Man (Ultimate)",
            "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/3/50/531771b4e8c60/portrait_fantastic.jpg",
          },
        ]
      }
    >
      <styled.main>
        <main
          className="sc-fzoiQi dJeHQt"
        >
          <CharacterList
            characters={
              Array [
                Object {
                  "id": 1011377,
                  "name": "Spider-Man (Takuya Yamashiro)",
                  "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/f/50/4be86ad8ada17/portrait_fantastic.jpg",
                },
                Object {
                  "id": 1011010,
                  "name": "Spider-Man (Ultimate)",
                  "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/3/50/531771b4e8c60/portrait_fantastic.jpg",
                },
              ]
            }
          >
            <styled.ul>
              <ul
                className="sc-AxgMl bVLvwi"
              >
                <styled.li
                  key="1011377"
                >
                  <li
                    className="sc-AxheI kdRVYU"
                  >
                    <CharacterCard
                      character={
                        Object {
                          "id": 1011377,
                          "name": "Spider-Man (Takuya Yamashiro)",
                          "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/f/50/4be86ad8ada17/portrait_fantastic.jpg",
                        }
                      }
                      isFaved={null}
                      onClick={[Function]}
                      onFaved={[Function]}
                      onUnfaved={[Function]}
                    >
                      <styled.section
                        alt="Spider-Man (Takuya Yamashiro)"
                        onClick={[Function]}
                      >
                        <section
                          alt="Spider-Man (Takuya Yamashiro)"
                          className="sc-AxirZ bCKUBB"
                          onClick={[Function]}
                        >
                          <styled.img
                            src="http://i.annihil.us/u/prod/marvel/i/mg/f/50/4be86ad8ada17/portrait_fantastic.jpg"
                          >
                            <img
                              className="sc-AxiKw jTQMHy"
                              src="http://i.annihil.us/u/prod/marvel/i/mg/f/50/4be86ad8ada17/portrait_fantastic.jpg"
                            />
                          </styled.img>
                          <Styled(Star)
                            isSolid={null}
                            onClick={[Function]}
                          >
                            <Star
                              className="sc-AxhUy iYWvJI"
                              isSolid={null}
                              onClick={[Function]}
                            >
                              <Styled(FontAwesomeIcon)
                                className="sc-AxhUy iYWvJI"
                                icon={
                                  Object {
                                    "icon": Array [
                                      576,
                                      512,
                                      Array [],
                                      "f005",
                                      "M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z",
                                    ],
                                    "iconName": "star",
                                    "prefix": "far",
                                  }
                                }
                                onClick={[Function]}
                              >
                                <FontAwesomeIcon
                                  border={false}
                                  className="sc-AxjAm jumYnU sc-AxhUy iYWvJI"
                                  fixedWidth={false}
                                  flip={null}
                                  icon={
                                    Object {
                                      "icon": Array [
                                        576,
                                        512,
                                        Array [],
                                        "f005",
                                        "M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z",
                                      ],
                                      "iconName": "star",
                                      "prefix": "far",
                                    }
                                  }
                                  inverse={false}
                                  listItem={false}
                                  mask={null}
                                  onClick={[Function]}
                                  pull={null}
                                  pulse={false}
                                  rotation={null}
                                  size={null}
                                  spin={false}
                                  swapOpacity={false}
                                  symbol={false}
                                  title=""
                                  transform={null}
                                >
                                  <svg
                                    aria-hidden="true"
                                    className="svg-inline--fa fa-star fa-w-18 sc-AxjAm jumYnU sc-AxhUy iYWvJI"
                                    data-icon="star"
                                    data-prefix="far"
                                    focusable="false"
                                    onClick={[Function]}
                                    role="img"
                                    style={Object {}}
                                    viewBox="0 0 576 512"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                                      fill="currentColor"
                                      style={Object {}}
                                    />
                                  </svg>
                                </FontAwesomeIcon>
                              </Styled(FontAwesomeIcon)>
                            </Star>
                          </Styled(Star)>
                          <styled.h2>
                            <h2
                              className="sc-AxhCb lbMrWD"
                            >
                               
                              Spider-Man (Takuya Yamashiro)
                               
                            </h2>
                          </styled.h2>
                        </section>
                      </styled.section>
                    </CharacterCard>
                  </li>
                </styled.li>
                <styled.li
                  key="1011010"
                >
                  <li
                    className="sc-AxheI kdRVYU"
                  >
                    <CharacterCard
                      character={
                        Object {
                          "id": 1011010,
                          "name": "Spider-Man (Ultimate)",
                          "thumbnail": "http://i.annihil.us/u/prod/marvel/i/mg/3/50/531771b4e8c60/portrait_fantastic.jpg",
                        }
                      }
                      isFaved={null}
                      onClick={[Function]}
                      onFaved={[Function]}
                      onUnfaved={[Function]}
                    >
                      <styled.section
                        alt="Spider-Man (Ultimate)"
                        onClick={[Function]}
                      >
                        <section
                          alt="Spider-Man (Ultimate)"
                          className="sc-AxirZ bCKUBB"
                          onClick={[Function]}
                        >
                          <styled.img
                            src="http://i.annihil.us/u/prod/marvel/i/mg/3/50/531771b4e8c60/portrait_fantastic.jpg"
                          >
                            <img
                              className="sc-AxiKw jTQMHy"
                              src="http://i.annihil.us/u/prod/marvel/i/mg/3/50/531771b4e8c60/portrait_fantastic.jpg"
                            />
                          </styled.img>
                          <Styled(Star)
                            isSolid={null}
                            onClick={[Function]}
                          >
                            <Star
                              className="sc-AxhUy iYWvJI"
                              isSolid={null}
                              onClick={[Function]}
                            >
                              <Styled(FontAwesomeIcon)
                                className="sc-AxhUy iYWvJI"
                                icon={
                                  Object {
                                    "icon": Array [
                                      576,
                                      512,
                                      Array [],
                                      "f005",
                                      "M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z",
                                    ],
                                    "iconName": "star",
                                    "prefix": "far",
                                  }
                                }
                                onClick={[Function]}
                              >
                                <FontAwesomeIcon
                                  border={false}
                                  className="sc-AxjAm jumYnU sc-AxhUy iYWvJI"
                                  fixedWidth={false}
                                  flip={null}
                                  icon={
                                    Object {
                                      "icon": Array [
                                        576,
                                        512,
                                        Array [],
                                        "f005",
                                        "M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z",
                                      ],
                                      "iconName": "star",
                                      "prefix": "far",
                                    }
                                  }
                                  inverse={false}
                                  listItem={false}
                                  mask={null}
                                  onClick={[Function]}
                                  pull={null}
                                  pulse={false}
                                  rotation={null}
                                  size={null}
                                  spin={false}
                                  swapOpacity={false}
                                  symbol={false}
                                  title=""
                                  transform={null}
                                >
                                  <svg
                                    aria-hidden="true"
                                    className="svg-inline--fa fa-star fa-w-18 sc-AxjAm jumYnU sc-AxhUy iYWvJI"
                                    data-icon="star"
                                    data-prefix="far"
                                    focusable="false"
                                    onClick={[Function]}
                                    role="img"
                                    style={Object {}}
                                    viewBox="0 0 576 512"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M528.1 171.5L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6zM388.6 312.3l23.7 138.4L288 385.4l-124.3 65.3 23.7-138.4-100.6-98 139-20.2 62.2-126 62.2 126 139 20.2-100.6 98z"
                                      fill="currentColor"
                                      style={Object {}}
                                    />
                                  </svg>
                                </FontAwesomeIcon>
                              </Styled(FontAwesomeIcon)>
                            </Star>
                          </Styled(Star)>
                          <styled.h2>
                            <h2
                              className="sc-AxhCb lbMrWD"
                            >
                               
                              Spider-Man (Ultimate)
                               
                            </h2>
                          </styled.h2>
                        </section>
                      </styled.section>
                    </CharacterCard>
                  </li>
                </styled.li>
              </ul>
            </styled.ul>
          </CharacterList>
          <CharacterModal />
        </main>
      </styled.main>
    </Home>
  </Router>
</MemoryRouter>
`;
